crossfire/__init__.py

```diff
@@ -51,16 +58,34 @@ def occurrences(
     )
 
 
+def flatten_df(row, column_name):
+    column_data = row[column_name]
+    return Series(
+        {f"{column_name}_{key}": value for key, value in column_data.items()}
+    )
+
+
 def flatten(data, nested_columns=None):
     nested_columns = set(nested_columns or NESTED_COLUMNS)
     if not nested_columns.issubset(NESTED_COLUMNS):
         raise NestedColumnError(nested_columns)
-    if not data:
+    if isinstance(data, DataFrame) and not data.empty:
+        keys = set(data.columns) & nested_columns
+        for key in keys:
+            data = pd_concat(
+                [
+                    data.drop(key, axis=1),
+                    data.apply(flatten_df, args=(key,), axis=1),
+                ],
+                axis=1,
+            )
+
         return data
-    if isinstance(data, list):
-        keys = set(data[0].keys()) & nested_columns
-        for item in data:
-            for key in keys:
-                item.update({f"{key}_{k}": v for k, v in item.get(key).items()})
-                item.pop(key)
+    elif not data:
         return data
```

Comment:
This should be an _early return_, not a late one.

Reply:
If you pass and empty `DataFrame`, what happens?

There is a bug there and an opportunity for optimization you are missing. 

An early return means returning before engaging in more demanding computing power, I/O or other resources. Before transforming stuff, you can check if it's an empty sequence (e.g., list) or an empty `DataFrame`.